File: java/connection/TypeDBSessionImpl.java
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.vaticle.typedb.driver.connection;

import com.vaticle.typedb.driver.api.TypeDBOptions;
import com.vaticle.typedb.driver.api.TypeDBSession;
import com.vaticle.typedb.driver.api.TypeDBTransaction;
import com.vaticle.typedb.driver.api.database.DatabaseManager;
import com.vaticle.typedb.driver.common.NativeObject;
import com.vaticle.typedb.driver.common.exception.TypeDBDriverException;

import java.util.ArrayList;
import java.util.List;

import static com.vaticle.typedb.driver.jni.typedb_driver.session_force_close;
import static com.vaticle.typedb.driver.jni.typedb_driver.session_get_database_name;
import static com.vaticle.typedb.driver.jni.typedb_driver.session_is_open;
import static com.vaticle.typedb.driver.jni.typedb_driver.session_new;
import static com.vaticle.typedb.driver.jni.typedb_driver.session_on_close;
import static com.vaticle.typedb.driver.jni.typedb_driver.session_on_reopen;

public class TypeDBSessionImpl extends NativeObject<com.vaticle.typedb.driver.jni.Session> implements TypeDBSession {
    private final Type type;
    private final TypeDBOptions options;

    private final List<SessionCallback> callbacks;

    TypeDBSessionImpl(DatabaseManager databaseManager, String database, Type type, TypeDBOptions options) {
        super(newNative(databaseManager, database, type, options));
        this.type = type;
        this.options = options;

        callbacks = new ArrayList<>();
    }

    private static com.vaticle.typedb.driver.jni.Session newNative(DatabaseManager databaseManager, String database, Type type, TypeDBOptions options) {
        try {
            return session_new(((TypeDBDatabaseManagerImpl)databaseManager).nativeObject, database, type.nativeObject, options.nativeObject);
        } catch (com.vaticle.typedb.driver.jni.Error e) {
            throw new TypeDBDriverException(e);
        }
    }

    @Override
    public boolean isOpen() {
        return session_is_open(nativeObject);
    }

    @Override
    public Type type() {
        return type;
    }

    @Override
    public String database_name() {
        return session_get_database_name(nativeObject);
    }

    @Override
    public TypeDBOptions options() {
        return options;
    }

    @Override
    public TypeDBTransaction transaction(TypeDBTransaction.Type type) {
        return transaction(type, new TypeDBOptions());
    }

    @Override
    public TypeDBTransaction transaction(TypeDBTransaction.Type type, TypeDBOptions options) {
        return new TypeDBTransactionImpl(this, type, options);
    }

    @Override
    public void onClose(Runnable function) {
        try {
            SessionCallback callback = new SessionCallback(function);
            callbacks.add(callback);
            session_on_close(nativeObject, callback.released());
        } catch (com.vaticle.typedb.driver.jni.Error error) {
            throw new TypeDBDriverException(error);
        }
    }

    @Override
    public void onReopen(Runnable function) {
        try {
            SessionCallback callback = new SessionCallback(function);
            callbacks.add(callback);
            session_on_reopen(nativeObject, callback.released());
        } catch (com.vaticle.typedb.driver.jni.Error error) {
            throw new TypeDBDriverException(error);
        }
    }

    @Override
    public void close() {
        try {
            session_force_close(nativeObject);
        } catch (com.vaticle.typedb.driver.jni.Error error) {
            throw new TypeDBDriverException(error);
        } finally {
            callbacks.clear();
        }
    }

    static class SessionCallback extends com.vaticle.typedb.driver.jni.SessionCallbackDirector {
        private final Runnable function;

        SessionCallback(Runnable function) throws com.vaticle.typedb.driver.jni.Error {
            this.function = function;
        }

        @Override
        public void callback() {
            function.run();
        }
    }
}

File: nodejs/connection/TypeDBSessionImpl.ts
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import {Database} from "../api/connection/database/Database";
import {TypeDBOptions} from "../api/connection/TypeDBOptions";
import {SessionType, TypeDBSession} from "../api/connection/TypeDBSession";
import {TransactionType, TypeDBTransaction} from "../api/connection/TypeDBTransaction";
import {ErrorMessage} from "../common/errors/ErrorMessage";
import {TypeDBDriverError} from "../common/errors/TypeDBDriverError";
import {RequestBuilder} from "../common/rpc/RequestBuilder";
import {TypeDBStub} from "../common/rpc/TypeDBStub";
import {RequestTransmitter} from "../stream/RequestTransmitter";
import {TypeDBTransactionImpl} from "./TypeDBTransactionImpl";
import {ServerDriver, TypeDBDriverImpl} from "./TypeDBDriverImpl";
import {TypeDBDatabaseImpl} from "./TypeDBDatabaseImpl";
import SESSION_CLOSED = ErrorMessage.Driver.SESSION_CLOSED;

export class TypeDBSessionImpl implements TypeDBSession {
    private readonly _driver: TypeDBDriverImpl;
    private readonly _databaseName: string;
    private readonly _type: SessionType;
    private readonly _options: TypeDBOptions;
    private _id: string;
    private _database: TypeDBDatabaseImpl;
    private _isOpen: boolean;
    private _serverDriver?: ServerDriver;
    private _pulse: NodeJS.Timeout;
    private _networkLatencyMillis: number;
    private _transactions: Set<TypeDBTransaction.Extended>;
    private _onClose: (() => void)[]
    private _onReopen: (() => void)[]

    constructor(databaseName: string, type: SessionType, options: TypeDBOptions, driver: TypeDBDriverImpl) {
        this._databaseName = databaseName;
        this._type = type;
        this._options = options;
        this._driver = driver;
        this._isOpen = false;
        this._transactions = new Set();
    }

    async open(): Promise<void> {
        this._database = await this._driver.databases.get(this._databaseName) as TypeDBDatabaseImpl;
        await this._database.runFailsafe(serverDriver => this.openAt(serverDriver));
    }

    private async reopenAt(serverDriver: ServerDriver): Promise<void> {
        await this.openAt(serverDriver);
        for (const callback of this._onReopen) {
            callback();
        }
    }

    private async openAt(serverDriver: ServerDriver): Promise<void> {
        const openReq = RequestBuilder.Session.openReq(this._databaseName, this._type.proto(), this._options.proto())

        const start = (new Date()).getMilliseconds();
        const res = await serverDriver.stub.sessionOpen(openReq);
        const end = (new Date()).getMilliseconds();
        this._networkLatencyMillis = Math.max((end - start) - res.server_duration_millis, 1);

        this._id = "0x" + Buffer.from(res.session_id).toString("hex");
        this._serverDriver = serverDriver;
        this._isOpen = true;
        this._pulse = setTimeout(() => this.pulse(), 5000);
    }

    public onClose(callback: () => void) {
        this._onClose.push(callback)
    }

    public onReopen(callback: () => void) {
        this._onReopen.push(callback)
    }

    async close(): Promise<void> {
        if (this._isOpen) {
            this._isOpen = false;
            for (const tx of this._transactions) {
                await tx.close();
            }
            for (const callback of this._onClose) {
                callback();
            }
            this._driver.closeSession(this);
            clearTimeout(this._pulse);
            const req = RequestBuilder.Session.closeReq(this._id);
            await this._serverDriver.stub.sessionClose(req);
        }
    }

    closed(transaction: TypeDBTransaction.Extended): void {
        this._transactions.delete(transaction);
    }

    async transaction(type: TransactionType, options?: TypeDBOptions): Promise<TypeDBTransaction> {
        if (!this.isOpen()) throw new TypeDBDriverError(SESSION_CLOSED);
        if (!options) options = new TypeDBOptions();
        return this._database.runFailsafe(async (serverDriver, _db, isFirstRun) => {
            if (!isFirstRun){
                await this.close();
                await this.reopenAt(serverDriver);
            }
            const transaction = new TypeDBTransactionImpl(this, type, options);
            await transaction.open();
            this._transactions.add(transaction);
            return transaction;
        });
    }

    get database(): Database {
        return this._database;
    }

    isOpen(): boolean {
        return this._isOpen;
    }

    get options(): TypeDBOptions {
        return this._options;
    }

    get type(): SessionType {
        return this._type;
    }

    get id() {
        return this._id;
    }

    get stub(): TypeDBStub {
        return this._serverDriver.stub;
    }

    get requestTransmitter(): RequestTransmitter {
        return this._serverDriver.transmitter;
    }

    get networkLatency() {
        return this._networkLatencyMillis;
    }

    private async pulse(): Promise<void> {
        if (!this._isOpen) return;
        const pulse = RequestBuilder.Session.pulseReq(this._id);
        try {
            const isAlive = await this._serverDriver.stub.sessionPulse(pulse);
            if (!isAlive) this._isOpen = false;
            else this._pulse = setTimeout(() => this.pulse(), 5000);
        } catch (e) {
            this._isOpen = false;
        }
    }
}
